src/object/ambient_sound.cpp

```diff
@@ -252,22 +126,94 @@ AmbientSound::collision(GameObject& other, const CollisionHit& hit_)
 void
 AmbientSound::draw(DrawingContext& context)
 {
-  if (Editor::is_active()) {
+  if (Editor::is_active())
     context.color().draw_filled_rect(m_col.m_bbox, Color(0.0f, 0.0f, 1.0f, 0.6f),
                                      0.0f, LAYER_OBJECTS);
-  }
 }
 
 void
 AmbientSound::stop_looping_sounds()
 {
-  stop_playing();
+  if (m_sound_source)
+    m_sound_source->stop(false);
 }
 
 void
 AmbientSound::play_looping_sounds()
 {
-  start_playing();
+  if (Editor::is_active())
+    return;
+
+  m_sound_source->play();
+}
+
+void
+AmbientSound::update(float dt_sec)
+{
+  if (Editor::is_active())
+    return;
+
+  const Player* const nearest_player = Sector::get().get_nearest_player(get_bbox().get_middle());
+  if (!nearest_player)
+    return;
+  const Rectf& player_bbox = nearest_player->get_bbox();
+  const Vector player_center = player_bbox.get_middle();
+
+  if (collision::intersects(player_bbox, get_bbox()))
+    m_sound_source->set_gain(m_volume);
+  else
+  {
+    float player_distance = m_radius+1;
+    if (player_center.x>=get_bbox().get_left() && player_center.x<=get_bbox().get_right())
+      player_distance = player_center.y<get_bbox().get_top() ? get_bbox().get_top()-player_center.y : player_center.y-get_bbox().get_bottom();
+    else if (player_center.y>=get_bbox().get_top() && player_center.y<=get_bbox().get_bottom())
+      player_distance = player_center.x<get_bbox().get_left() ? get_bbox().get_left()-player_center.x : player_center.x-get_bbox().get_right();
+    else if (player_center.x<=get_bbox().get_left() && player_center.y<=get_bbox().get_top())
+      player_distance = glm::distance(player_center, get_bbox().p1());
+    else if (player_center.x>=get_bbox().get_right() && player_center.y<=get_bbox().get_top())
+      player_distance = glm::distance(player_center, get_bbox().p1()+Vector(get_bbox().get_width(), 0));
+    else if (player_center.x<=get_bbox().get_left() && player_center.y>=get_bbox().get_bottom())
+      player_distance = glm::distance(player_center, get_bbox().p1()+Vector(0, get_bbox().get_height()));
+    else if (player_center.x>=get_bbox().get_right() && player_center.y>=get_bbox().get_bottom())
+      player_distance = glm::distance(player_center, get_bbox().p2());
+    m_sound_source->set_gain(std::max(m_radius_in_px-player_distance, 0.0f)/m_radius_in_px*m_volume);
```

Comment:
Spaces between comparators and operators like `<=` and `-` would be nice for readability.